File: php-bot/src/Services/SettingsService.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Config;
use App\Database;

/**
 * Runtime-configurable settings. Config::tradingDefaults() is the schema +
 * fallback; overrides live in the `settings` table so the Admin Panel can
 * change any knob without a redeploy. Cached for the lifetime of one PHP
 * request/cron run (there is no long-lived process to cache across).
 */
final class SettingsService
{
    /** @var array<string, mixed>|null */
    private static ?array $cache = null;

    /** @return array<string, mixed> */
    public static function getAll(bool $forceReload = false): array
    {
        if (self::$cache === null || $forceReload) {
            self::$cache = self::load();
        }
        return self::$cache;
    }

    /** @return array<string, mixed> */
    private static function load(): array
    {
        $merged = Config::tradingDefaults();
        $stmt = Database::connection()->query('SELECT `key`, value FROM settings');
        foreach ($stmt->fetchAll() as $row) {
            $key = $row['key'];
            if (array_key_exists($key, $merged)) {
                $merged[$key] = self::coerce($row['value'], $merged[$key]);
            }
        }
        return $merged;
    }

    private static function coerce(string $raw, mixed $default): mixed
    {
        if (is_bool($default)) {
            return in_array(strtolower(trim($raw)), ['1', 'true', 'yes', 'on'], true);
        }
        if (is_int($default)) {
            return (int) round((float) $raw);
        }
        if (is_float($default)) {
            return (float) $raw;
        }
        return $raw;
    }

    public static function get(string $key): mixed
    {
        return self::getAll()[$key] ?? null;
    }

    public static function set(string $key, mixed $value): void
    {
        $defaults = Config::tradingDefaults();
        if (!array_key_exists($key, $defaults)) {
            throw new \InvalidArgumentException("Unknown setting: {$key}");
        }
        $strValue = is_bool($value) ? ($value ? '1' : '0') : (string) $value;

        $pdo = Database::connection();
        $stmt = $pdo->prepare('SELECT id FROM settings WHERE `key` = :key');
        $stmt->execute(['key' => $key]);
        if ($stmt->fetch()) {
            $upd = $pdo->prepare('UPDATE settings SET value = :value, updated_at = :now WHERE `key` = :key');
            $upd->execute(['value' => $strValue, 'now' => Database::now(), 'key' => $key]);
        } else {
            $ins = $pdo->prepare('INSERT INTO settings (`key`, value, updated_at) VALUES (:key, :value, :now)');
            $ins->execute(['key' => $key, 'value' => $strValue, 'now' => Database::now()]);
        }
        self::getAll(true);
    }

    public static function valueType(string $key): string
    {
        return gettype(Config::tradingDefaults()[$key] ?? null);
    }

    /** @return string[] */
    public static function allKeys(): array
    {
        return array_keys(Config::tradingDefaults());
    }
}

File: app/services/settings_service.py
"""Runtime-configurable settings.

TRADING_DEFAULTS (config.py) is the schema + fallback value for every knob.
Overrides are persisted in the `settings` DB table so the Admin Panel can
change any of them without a restart or redeploy. Values are cached in
memory and reloaded whenever something is written.
"""
from __future__ import annotations

from typing import Any

from sqlalchemy import select

from app.core.database import get_session
from app.core.models import Setting
from config import TRADING_DEFAULTS

_cache: dict[str, Any] | None = None


def _coerce(raw: str, default: Any) -> Any:
    if isinstance(default, bool):
        return raw.strip().lower() in ("1", "true", "yes", "on")
    if isinstance(default, int):
        return int(float(raw))
    if isinstance(default, float):
        return float(raw)
    return raw


async def _load() -> dict[str, Any]:
    global _cache
    merged = dict(TRADING_DEFAULTS)
    async with get_session() as session:
        result = await session.execute(select(Setting))
        for row in result.scalars():
            if row.key in merged:
                try:
                    merged[row.key] = _coerce(row.value, merged[row.key])
                except (ValueError, TypeError):
                    pass  # keep default if a stored value became invalid
    _cache = merged
    return merged


async def get_all(force_reload: bool = False) -> dict[str, Any]:
    global _cache
    if _cache is None or force_reload:
        return await _load()
    return _cache


async def get(key: str) -> Any:
    cache = await get_all()
    return cache[key]


async def set(key: str, value: Any) -> None:
    if key not in TRADING_DEFAULTS:
        raise KeyError(f"Unknown setting: {key}")
    str_value = str(value)
    async with get_session() as session:
        result = await session.execute(select(Setting).where(Setting.key == key))
        row = result.scalar_one_or_none()
        if row is None:
            session.add(Setting(key=key, value=str_value))
        else:
            row.value = str_value
    await get_all(force_reload=True)


def value_type(key: str) -> type:
    return type(TRADING_DEFAULTS[key])


def all_keys() -> list[str]:
    return list(TRADING_DEFAULTS.keys())
